### PR TITLE
feat(sitemap): add support for "lastmod"

### LINKED ISSUE #2604: Add lastmod to sitemap
## 🐛 Bug Report

The XML sitemaps currently output `loc`, `changefreq` and `priority` for every `url` set. I would propose dropping the `changefreq` and `priority` fields, as none of the search engines use these, and instead _adding_ the `lastmod` field, with the last modification date of the file.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

## To Reproduce

(Write your steps here:)

1. Open any DocuSaurus v2 sitemap :)

## Expected behavior

The current output would be:

```xml
<url>
        <loc>https://developer.yoast.com/features/canonical-urls/api</loc>
        <changefreq>weekly</changefreq>
        <priority>0.5</priority>
</url>
```

(Write what you thought would happen.)

## Actual Behavior

I propose changing it to:

```xml
<url>
        <loc>https://developer.yoast.com/features/canonical-urls/api</loc>
        <lastmod>2020-04-14T11:22:05+00:00</lastmod>
</url>
```

## Your Environment

- Docusaurus version used: v2

### PR DESCRIPTION
**Pull request by slorber**
## Motivation

Fix https://github.com/facebook/docusaurus/issues/2604

Add new sitemap options:

```js
{
  lastmod: null | 'date' | 'datetime'
  priority: null,
  changefreq: null,
}
```

- Add a new sitemap option `lastmod` that permits to output `<lastmod>` tags in the sitemap, formatted as `YYYY-MM-DD` date, or ISO datetime
- Allow to disable useless `<changefreq>` and `<priority>` sitemap tags (Google doesn't use them) with `null` values

Important notes:
- `lastmod` only works for routes backed and requires plugin authors to add new route metadata when calling `addRoute()`
- All our official content plugins provide this route metadata
- If plugins allow to provide a last updated time as front matter, then it's possible for the sitemap plugin to use this explicit date in priority. Otherwise, a timestamp will be inferred from the Git history of the content source file.
- We are able to output a `<lastmod>` tag for React page components as well. However it will only use the React page source Git history, and will not check history of its dependencies.
- Not all routes will have a `<lastmod>` tag. If a third-party plugin does not provide the new metadata, the sitemap can't infer a proper value. According to Google recommendations, it's only relevant to use `<lastmod>` tags for routes backed by content, and not routes aggregating multiple documents (tags, blog pagination, blog archive, doc autogenerated index...)

## Test Plan

Unit tests + dogfood on our website

### Test links

- Website sitemap: https://deploy-preview-9954--docusaurus-2.netlify.app/sitemap.xml
- Sitemap options doc: https://deploy-preview-9954--docusaurus-2.netlify.app/docs/api/plugins/@docusaurus/plugin-sitemap/
- RouteMetadata doc: https://deploy-preview-9954--docusaurus-2.netlify.app/docs/api/plugin-methods/lifecycle-apis/#addRoute

## Related issues/PRs

Supersed PR https://github.com/facebook/docusaurus/pull/9234

### CODE REVIEW

**Review by johnnyreilly — COMMENTED** [review]
This is excellent work; I like that you're looking to make the default sitemap include `lastmod` by default with Docusaurus v4

File: packages/docusaurus-plugin-sitemap/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  'hourly',
+  'daily',
+  'weekly',
+  'monthly',
+  'yearly',
+  'always',
+  'never',
+] as const;
+
+export type ChangeFred = (typeof ChangeFreqList)[number];
```
[the remark this answers is not in the document]
Comment (johnnyreilly) [reply packages/docusaurus-plugin-sitemap/src/types.ts:25]: No. It's very important that Fred changes :laughing:

**Review by slorber — COMMENTED** [review]
LGTM 🎉 

Thanks for the review 🙏

**Comment by razzeee** [comment]
Is it expected, that lastmod is always today? (might be date of last deploy) 
For every page?

See https://docs.flathub.org/sitemap.xml
Code is here https://github.com/flathub-infra/documentation

**Comment by slorber** [comment]
@razzeee you need to configure your CI to fetch the full git history.
I may have missed to mention that in the plugin docs, but it should be there in docs/blog docs.

**Comment by johnnyreilly** [comment]
You may need:

```yml
      - name: Checkout 📥
        uses: actions/checkout@v4
        with:
          # Number of commits to fetch. 0 indicates all history for all branches and tags.
          fetch-depth: 0
          submodules: true
```

https://github.com/johnnyreilly/blog.johnnyreilly.com/blob/2a4665f3accd44f9199b8232310368dedac37f19/.github/workflows/build-and-deploy-static-web-app.yml#L33C1-L38C27

**Comment by razzeee** [comment]
Thanks, working now

**Comment by pmarschik** [comment]
I've now got around to trying the released version.
On our CI it works as expected, however locally I have the git config `log.showSignature = true` which causes `npm docusaurus build` to fail with:
```
[ERROR] Loading of version failed for version current

[ERROR] Error: Unable to build website for locale en.
[...]

Error: Can't process doc metadata for doc at path path=/Users/<redacted> in version name=current

**Comment by slorber** [comment]
Thanks @pmarschik , I hoped for both solutions at once in https://github.com/facebook/docusaurus/pull/10022

> and extend the regex with the tags and allow multiline matches.

Not sure what you mean here

**Comment by pmarschik** [comment]
> Thanks @pmarschik , I adopted both solutions at once in #10022
> 
> > and extend the regex with the tags and allow multiline matches.
> 
> Not sure what you mean here

@slorber Thanks for the quick fix.

I meant exactly what you merged in #10022. Except with [multiline](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/multiline#description) I meant the RegExp could be `/^RESULT:(?<timestamp>\d+),(?<author>.+)$/m` instead of `/(?:^|\n)RESULT:(?<timestamp>\d+),(?<author>.+)(?:$|\n)/`

**Pull request merged**